kafka/producer/sender.py

```diff
@@ -195,13 +195,18 @@ def _handle_produce_response(self, node_id, send_time, batches, response):
 
             for topic, partitions in response.topics:
                 for partition_info in partitions:
+                    error_message = None
                     if response.API_VERSION < 2:
                         partition, error_code, offset = partition_info
                         ts = None
-                    else:
+                    elif 2 <= response.API_VERSION <= 4:
                         partition, error_code, offset, ts = partition_info
+                    elif 5 <= response.API_VERSION <= 7:
+                        partition, error_code, offset, ts, log_start_offset = partition_info
+                    else:
+                        partition, error_code, offset, ts, log_start_offset, _, error_message = partition_info
                     tp = TopicPartition(topic, partition)
-                    error = Errors.for_code(error_code)
+                    error = error_message or Errors.for_code(error_code)
```

Comment:
so for `API_VERSION` > 7, this will be the raw `partition_info`... why is that? What is an example `partition_info`? Should it instead be called `partition_error_code`? And in that case, should it be:

```
error_code = partition_error_code or response_error_code
error = Errors.for_code(error_code)
```
?

Reply:
@jeffwidman From http://kafka.apache.org/protocol.html#api_versions, `error_message` is 
```
The global error message summarizing the common root cause of the records that caused the batch to be dropped
```
So from my understanding, the broker will summarize the all the error messages into one, and provide them for you. The actual distinct messages plus codes are in the penultimate dropped / ignored unpack value, just because i found no need for it once the actual cumulated error messages were available all into one string

Reply:
@jeffwidman The structure seems to be as such:
```
Produce Response (Version: 8) => [responses] throttle_time_ms 
  responses => topic [partition_responses] 
    topic => STRING
    partition_responses => partition error_code base_offset log_append_time log_start_offset [record_errors] error_message 
      partition => INT32
      error_code => INT16
      base_offset => INT64
      log_append_time => INT64
      log_start_offset => INT64
      record_errors => batch_index batch_index_error_message 
        batch_index => INT32
        batch_index_error_message => NULLABLE_STRING
      error_message => NULLABLE_STRING
  throttle_time_ms => INT32
```
So the last part in partition_info is a UTF-8 string expressing the combined error that happened broker side.
But this PR will change the error passed to `_complete_batch `, which will result in callbacks called with an incorrect object. This is a regression and will also (!) break handling for code in `_complete_batch `, as it checks for exact equality to ``Errors.NoError`` there.

Reply:
@gabriel-tincu Please pass the ``error_message`` explicitly using arguments up to the log in ``batch.done()``. There we can introduce it as part of the log statement, as you already did with ``log_start_offset``. I am not sure if we can safely change the error raised to the users (cc @dpkp ), but from how the code stands `Future.failure` has this magic line:
```python
    def failure(self, e):
        assert not self.is_done, 'Future is already complete'
        self.exception = e if type(e) is not type else e()        # <--------------
        assert isinstance(self.exception, BaseException), (
            'future failed without an exception')
        self.is_done = True
        self._call_backs('errback', self._errbacks, self.exception)
        return self
```
So we raise instances of exceptions to the users always. Thus if we change the
```python
self.produce_future.failure(exception)
```
line in ``batch.done()`` to something like:
```python
 if not error_message:
    exc = exception()
else:
    exc = exception(error_message)
self.produce_future.failure(exc)
```
We should be able to pass error message from broker to user without introducing any incompatible changes in code.

Reply:
@tvoinarovskyi ` Would keeping the old value be considered more stable then? I figured the cummulated strings would offer more information to the developer, but if one was to keep the old error implementation, there is no way to use that, since the api requires an error code as well, and there is no error code associated with the cumulated string

Reply:
Thanks @tvoinarovskyi I was feeling a little queasy over the hand-wavy way this was dealing with errors, but wasn't sure how what a good way was. Thanks for digging in and stepping through it.